### PR TITLE
DBConnection and Pooling

### PR DESCRIPTION
**Pull request by hamiltop**

### CODE REVIEW

File: lib/rethinkdb/connection/multiplexer.ex [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    |> GenServer.reply({token, msg})
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, _ref, :process, _pid, reason}, state) do
+    {:stop, reason, state}
+  end
+
+  @doc false
+  def recv_loop(pid, {transport, sock} = conn) do
```
Comment (hamiltop) [inline lib/rethinkdb/connection/multiplexer.ex]: Why have this in a separate process? You're basically reimplementing `active: :once`.

**Comment by almightycouch** [comment]
Oops, my code is using the `with` keyword from Elixir 1.3. Which doesn't compile with your Travis-CI integration using version 1.1.

File: lib/rethinkdb/connection/multiplexer.ex [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    with {:ok, sock} <- transport.connect(String.to_char_list(host), port, sock_opts),
+         {:ok, conn} <- init_connection(transport, sock),
+         {:ok, _nil} <- handshake(conn, user, pass),
+         {_pid, ref} <- spawn_monitor(__MODULE__, :recv_loop, [self, conn]),
+    do:  {:ok, %__MODULE__{conn: conn, ets: :ets.new(__MODULE__, [:set, :private]), ref: ref}}
+  end
+
+  def handle_call({:send, nil, msg}, from, %__MODULE__{conn: conn, token: token, ets: table} = state) do
+    :ok = send_payload(conn, token, msg)
+    :ets.insert(table, {token, from})
```
Comment (hamiltop) [inline lib/rethinkdb/connection/multiplexer.ex]: Why use `:ets`? Why not just use the process state? The advantage of `:ets` is you can have readers from outside of this process (which we do not have).

File: lib/rethinkdb/connection.ex [hunk 3]
```diff
@@ -123,213 +119,101 @@ defmodule RethinkDB.Connection do
...
+  #
+  # DBConnection API
+  #
 
-  def handle_call({:stop, token}, from, state) do
-    query = "[3]"
-    Request.make_request(query, token, from, state)
+  def connect(options) do
+    # Starts the multiplexer process and stores its pid and some default options to the state.
+    {:ok, pid} = Multiplexer.start_link(options)
```
Comment (hamiltop) [inline lib/rethinkdb/connection.ex]: Why do we need a separate multiplexer process?

**Comment by hamiltop** [comment]
Go ahead and update travis.yml to use 1.3.

I don't think it's an unreasonable ask.

File: lib/rethinkdb/connection/multiplexer.ex (hunk 2 above)
Comment (almightycouch) [inline lib/rethinkdb/connection/multiplexer.ex]: Right. Using the process state to store responses in a `Map` may be a better choice here.
I thought that `:ets` tables would provide better performances for storing large data (changefeeds).

I'm going to refactor this.

File: lib/rethinkdb/connection.ex (hunk 3 above)
Comment (almightycouch) [inline lib/rethinkdb/connection.ex]: I found that splitting up the `DBConnection` behaviour and the `Multiplexer` into two separate processes (with their own state and mailbox) could be a good thing.

The main reason is that `Multiplexer.send_recv/3` blocks until a reply is received (using `{:noreply, state}` and `GenServer.reply/1`on the server side).

File: lib/rethinkdb/connection/multiplexer.ex (hunk 1 above)
Comment (almightycouch) [inline lib/rethinkdb/connection/multiplexer.ex]: 👍 I've missed that feature. Will refactor that.

File: lib/rethinkdb/connection.ex [hunk 4]
```diff
@@ -123,213 +119,101 @@ defmodule RethinkDB.Connection do
...
+        # to support default options when encoding the query.
+        message = describe(query, options)
+        |> Map.fetch!(:message)
+      else
+        message
+      end
 
-  def terminate(_reason, _state) do
-    :ok
+    case Multiplexer.send_recv(pid, message, options) do
```
Comment (hamiltop) [inline lib/rethinkdb/connection.ex]: Is this blocking the DBConnection process? Doesn't that prevent multiplexing?

It looks like calling `next` will work concurrently, but no other queries will run side by side.

File: lib/rethinkdb/connection/multiplexer.ex (hunk 1 above)
Comment (hamiltop) [inline lib/rethinkdb/connection/multiplexer.ex]: Take a look at my original code https://github.com/hamiltop/rethinkdb-elixir/blob/master/lib/rethinkdb/connection/request.ex#L24 . It's a bit complex because you don't how much data you will be receiving so you have to buffer in the process state.

File: lib/rethinkdb/connection.ex (hunk 4 above)
Comment (almightycouch) [inline lib/rethinkdb/connection.ex]: It does not prevent multiplexing because the `Multiplexer` process has it own request/reply loop. Mapping calling process ids with RethinkDB tokens.

So multiple clients can have concurrent queries running at the same time.

Comment (hamiltop) [inline lib/rethinkdb/connection.ex]: The multiplexer will do multiple requests, sure (it looks like calling `next` works perfectly fine). But will the Connection allow multiple requests?

It looks to me like this is the current flow:

1. User calls `run` which calls `DBConnection.execute!`
2. In connection process, `handle_execute` is called.
3. `handle_execute` calls `Multiplexer.send_recv` which blocks.
4. connection process is now blocked until Multiplexer.send_recv returns.

Comment (almightycouch) [inline lib/rethinkdb/connection.ex]: The [`handle_execute/4`](https://hexdocs.pm/db_connection/DBConnection.html#c:handle_execute/4) callback is  called in the client process. Which means that the connection can handle concurrent queries from multiple processes.

I wrote a small test:
https://gist.github.com/almightycouch/6fed5356c60a6ef70a41eea31777fa62

When i log the output inside `send_recv/4` with following line:
```elixir
IO.inspect "send_recv from #{inspect self} to #{inspect pid}"
```

this is my output:

```
"send_recv from #PID<0.178.0> to #PID<0.176.0>"
"send_recv from #PID<0.177.0> to #PID<0.176.0>"
"send_recv from #PID<0.179.0> to #PID<0.176.0>"
```

Using`:poolboy`:

```elixir
{:ok, pid} = RethinkDB.Connection.start_link, pool: DBConnection.Poolboy
```

this is my output:

```
"send_recv from #PID<0.292.0> to #PID<0.247.0>"
"send_recv from #PID<0.290.0> to #PID<0.251.0>"
"send_recv from #PID<0.291.0> to #PID<0.249.0>"
```

Which is i think what is excepted.

Comment (hamiltop) [inline lib/rethinkdb/connection.ex]: That code works because it has one process calling `run` and 3 calling `next` (via feed Enumeration). `next` bypasses DBConnection and hits multiplexer directly.

However, I do think you are right. I did not realize `handle_execute` is called from the client. That should be fine then. Adapting the test for multiple producers would likely show the correct behavior, but I'd love to make sure.

Comment (almightycouch) [inline lib/rethinkdb/connection.ex]: I do 100% understand your worries :fire:. `DBConnection` works completely differently than `Connection`.
I'm also pretty sure it should work that way. The tests are running with `async: true` and do run successfully.

In the above example (with `:poolboy`), the producer also hits a different connection process each time:

```
"send_recv from #PID<0.79.0> to #PID<0.178.0>"
"send_recv from #PID<0.79.0> to #PID<0.192.0>"
"send_recv from #PID<0.79.0> to #PID<0.194.0>"
```

Comment (hamiltop) [inline lib/rethinkdb/connection.ex]: The specific case I'm worried about hasn't been covered here. Specifically, multiple producers on a single connection via `execute`. I wanted to make sure DBConnection wasn't enforcing mutual exclusion. I dove into the code though and tweaked your test and it's all fine. Multiplexing in producers works fine.

Thanks for bearing with me on this.

**Comment by almightycouch** [comment]
Are there any benefits for using `[active: :once]` over `[active: true]`? If we poll every time we receive a message.

**Comment by hamiltop** [comment]
Understanding how TCP works is important here.

Assume we used `active: true`. The db sends data to us over TCP. The OS networking layer receives that data, informs the OS process listening on that socket (BEAM) that data is available and buffers the data. BEAM then accepts the data (clearing the OS buffer) and sends it to our process and the data is buffered in the process mailbox. What happens if we can't process the data fast enough? The process mailbox grows.

Now let's look at `active: :once`. The db sends data to us over TCP. The OS networking layer receives that data, informs the OS process listening on that socket (BEAM) that data is available and buffers the data. BEAM now checks to see if the process owning that socket is `active`. If it is, it receives the data (clearing the OS buffer) and sends it to that process. if it is not active, it lets the data sit in the OS buffer. If the OS buffer is full, the OS will stop acking packets over TCP, and the DB will stop sending them. Which is exactly how TCP is supposed to work.

TLDR `active: :once` provides back pressure.

**Comment by almightycouch** [comment]
That is a really good explanation 👍.

I was wondering as I'm trying to implement a `GenStage` for a Twitter stream in an other project.
Did not know that `active: true` would result in back pressuring at the TCP level. This is really nice!

**Comment by hamiltop** [comment]
> Did not know that active: true would result in back pressuring at the TCP level. This is really nice!

Just to be clear `active: :once` back pressures at the TCP level.

**Comment by almightycouch** [comment]
Sorry, I meant to write `:once` 😅.

**Comment by hamiltop** [comment]
@almightycouch What's needed to get this in a mergeable state?

**Comment by almightycouch** [comment]
* Fix/modify failing, missing  connection tests. Check (`DBConnection` backoff options and pool specific options).
* Add  tests for `:poolboy` and `:sbroker` (I've nevers tested the latter).
* Query responses, especially errors should work as excepted. Not sure if this is always the case with the new `{:ok, rep}`, `{:error, error}` API. 

I'm currently traveling through eastern Europe and will be back home in september. Feel free to make changes and refactor in the mean time.

**Comment by almightycouch** [comment]
* Fix/modify failing, missing  connection tests. Check (`DBConnection` backoff options and Pool specific options).
* Add  tests for `:poolboy` and `:sbroker` (I've nevers tested the latter).
* Query responses, especially errors should work as excepted. Not sure if this is always the case with the new `{:ok, rep}`, `{:error, error}` API. 

I'm currently traveling through eastern Europe and will be back home in september. Feel free to make changes and refactor in the mean time.

**Comment by almightycouch** [comment]
Also

> By default the DBConnection provides a single connection. However the :pool option can be set to use a pool of connections. If a pool is used the module must be passed as an option - unless inside a run/3 or transaction/3 fun and using the run/transaction connection reference (t).

I think, my implementation is a unpleasant hack:

```elixir
def __pool__(conn) do
    info =
      unless is_pid(conn) do
        Process.whereis(conn)
      else
        conn
      end
      |> Process.info(:dictionary)
    {pool, _, _} = elem(info, 1)[:"$initial_call"]
    case pool do
      :poolboy ->
        DBConnection.Poolboy
      _ ->
        DBConnection.Connection
    end
  end
```

I did not really understand the necessity to pass the `:pool` option for each request nor how this should be implemented properly.

**Comment by fishcakez** [comment]
I don't think `DBConnection` is a good fit for multiplexing. It is designed for simplex (SQL) connections for Ecto drivers. It's goals don't really extend beyond that and it forces particular decisions to get the desired Ecto behaviour for SQL adapters.

Before `DBConnection` we had a poolboy pool of supervising processes (not `Supervisor`s) that started and looked after connection processes and held information about an active transaction. This caused many problems for us because poolboy, the "supervisor", the connection and the calling process held state data of the connection. Unfortunately only the connection process (e.g. multiplexer) knows the state of the connection but it does not let the other processes know of this state because of the top down design. Therefore other processes can get out of sync and errors can be difficult to handle cleanly.

By starting the multiplexer in the connection process and passing its pid as the "socket" this PR returns to the pattern Ecto was going away from. That isn't to say you can't solve these problems, and the main issue for Ecto SQL was transaction handling that rethinkdb does not do.

To solve our issues with transaction status we separated transactions from pooling by passing the underlying socket/state to the calling process (e.g. process calling `Rethinkdb.Connection.run`). This meant that a calling process hods the full state and if it exits while executing then the socket must be closed because the state is lost. Therefore if a calling process exits while calling `Rethinkdb.Connection.run` the multiplexer will be closed, closing the cursors/change feeds (sorry I am not familiar with rethinkdb jargon) for other processes.

Because the state is in the calling process for the duration of a call, the connection process will ignore messages because it does not have the state to update. This means that if the connections process is a supervisor for the real connection process (e.g. multiplexer) and it traps exits it can miss an `:EXIT` message for any processes it start links to if it arrives while checked out. If the supervising process does not trap exits then it will exit if the real connection does. However in this situation the complex backoff/reconnect code provided by `DBConnection`.

The other main feature of `DBConnection` connection processes is that a queue is provided so that one calling process can use the socket at the time. When multiplexing this is not required and perhaps counter productive, unless the back pressure is desired for setting up cursors.

`DBConnection` also provides logging. However when going directly to the multiplexer, i.e. using a cursor, `DBConnection` is not used so there isn't logging. Meaning logging needs to be (re)implemented if it is desired.

The configuration, e.g. pool choice, is designed with Ecto in mind where configuration is fetched as a keyword from `Application.get_env` and it makes it easy to pass straight through to all layers: pool, transactions and connection. The current method of pool module selection is not suitable unless limiting the available pools, and unlikely to work well with the `DBConnection.Sojourn` pool where the initial call will depend on the broker callback module.

However limiting to `DBConnection.Connection` and `DBConnection.Poolboy` would work with the `__pool__` "hack". But given that and my previous points I am not sure there is enough benefit to using `DBConnection` over `:poolboy` just with `Connection` processes underneath.

**Comment by fishcakez** [comment]
Sorry to be negative about the PR and not to suggest an alternative - I am not aware of one designed for your needs. I don't think `DBConnection` is a good fit here because it makes trade offs that don't fit your use case. I think it would be better in the long run to write your own pool - even if that piggy backs on another library.  Of course if DBConnection is enough, perhaps only in the short term, then go for it.

**Comment by hamiltop** [comment]
@fishcakez Thanks for chiming in. I've mentioned this before but getting your feedback is one of my favorite parts of working on this driver. Keep it coming!

Gotta think about this for a bit.

**Comment by almightycouch** [comment]
Hello @hamiltop, @fishcakez.

I've been traveling through eastern Europe for the past 2 month. Did not find time to focus on this discussion. Sorry for that.

> Sorry to be negative about the PR and not to suggest an alternative - I am not aware of one designed for your needs. I don't think `DBConnection` is a good fit here because it makes trade offs that don't fit your use case. I think it would be better in the long run to write your own pool - even if that piggy backs on another library. Of course if `DBConnection` is enough, perhaps only in the short term, then go for it.

First thank you for the great explanation about how and why `DBConnection` came to life, and why it might not fit with the RethinkDB adapter.

The main reason for me to rewrite `RethinkDB.Connection` was to better integrate with Ecto. I knew that some specific features such as transactions and prepare/execute would not be required at all by the adapter. On the other side, I thought `RethinkDB.Connection` would profit a lot from features such as pooling, encode/decode without blocking the connection and exponential backoff to reconnect.

In the end, I think you are right and using `DBConnection` might bring more trade-offs than benefits.

* RethinkDB does not support transactions nor prepare/execute.
* `RethinkDB.Connection` does not depend on configuration (`Application.get_env`), having to use the `__pool__` hack might be a really bad thing in this case (overhead for each query).
* Having a pool of multiplexed connections might be the wrong thing for the reasons you listed above. Especially with `changes()` and crashes.

@hamiltop, sorry for having you putting that much effort/time reviewing code. It helped me to improve my knowledge on OTP, back-pressure, etc. 👍. I did not want to push any decisions in bad/wrong directions. Again, sorry for that.

Parts of the code might still be helpful, [RethinkDB 2.3's user authentication](https://www.rethinkdb.com/docs/permissions-and-accounts/) for example. Feel free to use it if you want to.

**Comment by hamiltop** [comment]
No worries. Getting code review in a setting like this is a fantastic way to share knowledge.

If anything, I'm happy someone else knows the codebase now. Please stick around and help contribute! I'd welcome a PR of just the user auth stuff.

**Comment by almightycouch** [comment]
> If anything, I'm happy someone else knows the codebase now.

😄

> Please stick around and help contribute! I'd welcome a PR of just the user auth stuff.

I'm a little busy right now. Will work on that as soon as possible.

**Comment by hamiltop** [comment]
Alright. I'm going to close this. Thanks for opening the user auth PR!
